service/utils/file.py

```diff
@@ -0,0 +1,222 @@
+                        # so we force empty values to None so they will appear as null
+                        # in the end JSON output { "a": null }
+                        if not value:
+                            row.update({key: None})
+
+                        # pull out requested csv headers if header is requested and header exists or add all the headers
+                        if (self.config.get("h") and key in self.config.get("h")) or (
+                            not self.config.get("h")
+                        ):
+                            rl[key] = row[key]
+
+                        # if prune is set, remove any columns with a value of None
+                        if self.config.get("prune") and row[key] is None and key in rl:
+                            del rl[key]
+
+                    # if prune is set and we are sorting, we have to ensure the rows have all the columns to sort on
+                    if self.config.get("prune") and self.config.get("s"):
+                        cur_kz = list(rl.keys())
+                        srt_on = list(map(itemgetter("column"), self.config.get("s")))
+                        if all(item in cur_kz for item in srt_on):
+                            dl.append(rl)
+                    else:
+                        dl.append(rl)
+
+                self.total_data = len(dl)
+
+        except TypeError:
+            raise TypeError("Invalid Universe File")
+
+        return self.filter_data(dl)
+
+    def sort_results(self, results, srt_ordr=None):
+        """
+        Custom sorting function that allows for sorting None while also maintaining locale aware sorting
+        config["nulls"] will allow for sorting None and putting them at the front or end of the list
+
+        l = [1, 3, 2, 5, 4, None, 7]
+        print('Last = ', sorted(l, key=lambda x: (x is None, x)))
+        Last = [1, 2, 3, 4, 5, 7, None]
+        print('First = ', sorted(l, key=lambda x: (x is not None, x)))
+        First = [None, 1, 2, 3, 4, 5, 7]
+
+        :param results: (list) e.g. [{'name': 'richard jones (earth-616)', 'appearances': 590, 'year': 1962}]
+        :param srt_ordr: (dict) e.g. {'column': 'name', 'sort': False}
+        :return: (tuple) e.g. (False, True, 1969) tuples are sorted by item type, this means that all non-None elements
+        will come first (since False < True), and then be sorted by value.
+        """
+        srt_dict = srt_ordr if srt_ordr is not None else self.config.get("s")
+
+        for i in reversed(srt_dict):
+            results.sort(key=lambda itm: self.sort_i18n_str(itm, i), reverse=i["sort"])
+
+        return results
+
+    def sort_i18n_str(self, row, direction):
+        itm = row[direction["column"]]
+
+        if itm is not None and isinstance(itm, str):
+            itm = locale.strxfrm(row[direction["column"]])
+
+        if self.config.get("nulls") == "first" and not self.config.get("prune"):
```

Comment:
This sorts incorrectly when sorting on a column with asc|desc - nulls have to survive the initial column sort